src/vtarget-out/inv-syn/inv_abc_parse.cc

```diff
@@ -0,0 +1,639 @@
+            }
+            else if (ref_val_name.find("committed_inst") == 0) {
+              //cube_has_abnormal_var = true; 
+              literal = (neg ? "~" : "") +  std::string("1'b1");
+              remove_this_cube = remove_this_cube ||  neg; // if it is "& ~1'b1" this cube is 0
+              // literal = "1'b1"; // void the literal if abnormal
+              // over-approximate it
+            } else if (ref_val_name.find(".") != ref_val_name.npos) {
+              // keep it
+              literal = (neg ? "~" : "") +  handle_range_ref(ref_val_name);
+              remove_this_cube = true; // remove this cube (we don't know what to do actually)
+            } else 
+              remove_this_cube = true;
+          }else {
+            literal = (neg ? "~" : "") +  handle_range_ref(ref_val_name);
+          }
+          /*
+          if(replace_outside_var_ref && ref_val_name != "__all_assumed_reg__[0:0]") {
+            std::string new_var_name = ReplaceAll(ReplaceAll(ReplaceAll(ref_val_name, "." , "_dot_"), "[", "_"), "]","_");
+            outside_reference.insert(new_var_name);
+            literal = (neg ? "~" : "") +  new_var_name;
+          } else // o.w. it is okay
+            literal = (neg ? "~" : "") +  handle_range_ref( ref_val_name );
+          */
+        } // refering to outside var
+        if(cube.empty())
+          cube = literal;
+        else
+          cube = cube + " & "  + literal;
+      } // end of a cube
+      // cube = "( " +  cube + ") "
+
+      // void the cube if abnormal
+      if( remove_this_cube) // cube_has_abnormal_var ||
+        cube = "1'b0" ; //cube = "1'b0"; // under-approximate it : total under-approximate
+      else {// if at least the cube has a normal var, it is a normal cube
+        has_a_normal_cube = true;
+        // only if this cube is not removed, we will add it 
+        if (parse_result.empty())
+          parse_result =  "(" + cube + ")";
+        else
+          parse_result = parse_result + " | (" + cube + ")";
+      }
+    } // deal with a line
+    ILA_ERROR_IF(has_a_normal_cube == false) << "No normal is deduced.";
+    ILA_ERROR_IF(parse_result.empty()) << "No invariants are extracted!";
+    if (!has_a_normal_cube || parse_result.empty() ) {
+      parse_result = "";
+      parse_succeed = false;
+    }
+    
+    parse_result = "~("  + parse_result + ")";
+    parse_result = ReplaceAll(parse_result, "__all_assumed_reg__[0:0]", "1'b0");
+  } // parse abc file
+}
+
+
+
+void AbcInvariantParser::parse(
+    const std::string & blif_name,
+    const std::string & abc_result_fn,
+    const std::string & gla_map_fn) {
+  
+  std::vector<std::string> blif_state_order;
+  std::vector<unsigned> blif_state_init;
+  { // read blif
+    std::ifstream fin(blif_name);
+    if (! fin.is_open()) {
+      ILA_ERROR << "Cannot read from: " <<blif_name;
+      return;
+    }
+    std::string line;
+    while(std::getline(fin,line)) {
+      if (line.find(".latch ") == 0) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Inefficient usage of string::find() in condition; string::starts_with() could be faster.](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)